[SFU] `ERR_CONNECTION_RESET` when getting live camera images

## Summary

Errors with the message `connection reset by peer` are encountered when executing `[p]sfu cam <name>`.

## Traceback

```log
[2022-11-08 23:32:03] [ERROR] red: Exception in command 'sfu cam'
Traceback (most recent call last):
  File "$PYTHON_PATH/site-packages/urllib3/connectionpool.py", line 665, in urlopen
    httplib_response = self._make_request(
  File "$PYTHON_PATH/site-packages/urllib3/connectionpool.py", line 421, in _make_request
    six.raise_from(e, None)
  File "<string>", line 3, in raise_from
  File "$PYTHON_PATH/site-packages/urllib3/connectionpool.py", line 416, in _make_request
    httplib_response = conn.getresponse()
  File "$PYTHON_LIB_PATH/http/client.py", line 1347, in getresponse
    response.begin()
  File "$PYTHON_LIB_PATH/http/client.py", line 307, in begin
    version, status, reason = self._read_status()
  File "$PYTHON_LIB_PATH/http/client.py", line 268, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "$PYTHON_LIB_PATH/socket.py", line 669, in readinto
    return self._sock.recv_into(b)
ConnectionResetError: [Errno 104] Connection reset by peer

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "$PYTHON_PATH/site-packages/requests/adapters.py", line 439, in send
    resp = conn.urlopen(
  File "$PYTHON_PATH/site-packages/urllib3/connectionpool.py", line 719, in urlopen
    retries = retries.increment(
  File "$PYTHON_PATH/site-packages/urllib3/util/retry.py", line 400, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "$PYTHON_PATH/site-packages/urllib3/packages/six.py", line 734, in reraise
    raise value.with_traceback(tb)
  File "$PYTHON_PATH/site-packages/urllib3/connectionpool.py", line 665, in urlopen
    httplib_response = self._make_request(
  File "$PYTHON_PATH/site-packages/urllib3/connectionpool.py", line 421, in _make_request
    six.raise_from(e, None)
  File "<string>", line 3, in raise_from
  File "$PYTHON_PATH/site-packages/urllib3/connectionpool.py", line 416, in _make_request
    httplib_response = conn.getresponse()
  File "$PYTHON_LIB_PATH/http/client.py", line 1347, in getresponse
    response.begin()
  File "$PYTHON_LIB_PATH/http/client.py", line 307, in begin
    version, status, reason = self._read_status()
  File "$PYTHON_LIB_PATH/http/client.py", line 268, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "$PYTHON_LIB_PATH/socket.py", line 669, in readinto
    return self._sock.recv_into(b)
urllib3.exceptions.ProtocolError: ('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "$PYTHON_PATH/site-packages/discord.py-1.7.3-py3.8.egg/discord/ext/commands/core.py", line 85, in wrapped
    ret = await coro(*args, **kwargs)
  File "/home/ubuntu/github/renchan/Ren/cogs/sfu/roads.py", line 119, in cam
    fetchedData = requests.get(camera, headers=self.headers)
  File "$PYTHON_PATH/site-packages/requests/api.py", line 75, in get
    return request('get', url, params=params, **kwargs)
  File "$PYTHON_PATH/site-packages/requests/api.py", line 60, in request
    return session.request(method=method, url=url, **kwargs)
  File "$PYTHON_PATH/site-packages/requests/sessions.py", line 533, in request
    resp = self.send(prep, **send_kwargs)
  File "$PYTHON_PATH/site-packages/requests/sessions.py", line 646, in send
    r = adapter.send(request, **kwargs)
  File "$PYTHON_PATH/site-packages/requests/adapters.py", line 498, in send
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))
```

## Reproduction

Execute `[p]sfu cam <name>` (`<name>` being any camera name from `[p]help sfu cam`) multiple times. The exact cause is unsure, but it can be intermittent, so reproduction may not be easy.

## Diagnosis

1. Tried modifying the URLs for getting camera images (setting `nocache` to `1`, removing other query args) and the error still occurred.
2. Tried changing the user agent from `"Mozilla/5.0"` to `"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"` (a very common user agent) and the error still occurred.
3. Tried doing HTTP requests with retry (with backoff factor 0.25) and the error seemed to occur significantly more rarely.

## Temporary solution

Apply retry with backoff on live camera image requests.

--- a/cogs/sfu/roads.py
+++ b/cogs/sfu/roads.py
@@ -7,6 +7,7 @@ from io import BytesIO
 import datetime
 import json
 import requests
+import requests.adapters
 from bs4 import BeautifulSoup
 import discord
 from redbot.core import commands
@@ -87,6 +88,17 @@ class SFURoads(SFUBase):
         self.sfuGroup.add_command(self.cam)
         self.sfuGroup.add_command(self.report)
 
+        # Retry with exponential backoff for HTTP requests
+        # (https://github.com/SFUAnime/Ren/issues/590)
+        retryBackoff = requests.adapters.Retry(total=5, backoff_factor=0.25)
+        httpRetryHandler = requests.adapters.HTTPAdapter(max_retries=retryBackoff)
+
+        self.httpRetrySession = requests.Session()
+        self.httpRetrySession.mount("http://", httpRetryHandler)
+
+    def cog_unload(self):
+        self.httpRetrySession.close()
+
     @commands.command()
     @commands.guild_only()
     async def cam(self, ctx: Context, cam: str = ""):
@@ -116,12 +128,14 @@ class SFURoads(SFUBase):
             return
 
         try:
-            fetchedData = requests.get(camera, headers=self.headers)
+            fetchedData = self.httpRetrySession.get(camera, headers=self.headers)
             fetchedData.raise_for_status()
         except requests.exceptions.HTTPError:
             await ctx.send(":warning: This webcam is currently unavailable!")
             # self.logger.error(exc_info=True)
             return
+        except requests.adapters.MaxRetryError:
+            await ctx.send(":warning: Unable to retrieve webcam image. Please try again.")
 
         if not fetchedData.content:
             # Make sure we don't fetch a zero byte file

